docs/developing.rst

```diff
@@ -742,3 +742,246 @@ Note the call to :func:`toil.job.Job.encapsulate` creates the
+
+   $ cd my_project
+   $ ls
+   userScript.py utilities.py
+   $ ./userScript.py …
+
+where ``userScript.py`` imports additional functionality from ``utilities.py``.
+Toil detects that userScript.py has sibling modules—independent of whether they
+are actually imported by the user script—and copies those to the workers as
+well before invoking jobs. This is a decent way of organizing the source code
+of reasonably complicated workflows.
+
+
+Hot-deploying a package hierarchy
+---------------------------------
+
+Remember that in Python, a package is a directory containing one or more
+``.py`` files, one of which must be called ``__init__.py``, and optionally
+other packages. For more involved workflows that contain a significant amount
+of code, this is the recommended way of organizing the source. Because we use a
+package hierarchy, we can't really refer to the user script as such, we now
+call it the *user module*. It's simply one of the modules in the package
+hierarchy and we need to inform Toil about that by invoking it with Python `-m`
+option. That way Toil will be able to identify the entire set of modules
+belonging to the workflow and copy the right ones to each worker. Let's say we
+organized our sample project as follows::
+
+   $ cd my_project
+   $ tree
+   .
+   ├── bar
+   │   ├── __init__.py
+   │   └── sub
+   │       ├── __init__.py
+   │       └── bla.py
+   └── foo
+       ├── __init__.py
+       └── main.py
+   
+   3 directories, 5 files
+   $ python -m foo.main
+
+Here the user module ``main.py`` does not reside in the current directory, but
+is part of a package called ``bar``, in a subdirectory of the current
+directory. Additional functionality is in a separate module called
+``bar.sub.bla`` which corresponds to ``bar/sub/bla.py``. Since we invoke the
+user module via ``python -m foo.main``, Toil can determine the root directory
+of the hierarchy, ``my_project`` in this case and send all Python modules
+underneath it to each worker. The ``-m`` option is documented `here`_
+
+.. _here: https://docs.python.org/2/using/cmdline.html#cmdoption-m
+
+When passing ``-m``, Python adds the current working directory to ``sys.path``,
+the list of root directories that are considered when resolving a module name
+like ``foo.main``. Without that convenience we'd have to run the workflow as
+``PYTHONPATH="$PWD" python -m foo.main``. Note that Toil can detect the root
+directory of the user module's package hierarchy even if it isn't the current
+working directory. In other words we could also do this::
+
+   $ cd my_project
+   $ export PYTHONPATH="$PWD"
+   $ cd /some/other/dir
+   $ python -m foo.main
+
+
+Hot-deploying a virtualenv
+--------------------------
+
+So far we've looked at running an isolated user script, a user script in
+conjunction with sibling modules and a user module that is part of an entire
+package tree. But what if our workflow depends on functionality from external
```

Comment:
Instead of:

> But what if our workflow depends on functionality from external dependencies, like those that can be downloaded from PyPI and installed via pip or easy_install?

Prefer:


> But what if our workflow requires external dependencies that can be downloaded from PyPI and installed via pip or easy_install?